cardano-api/src/Cardano/Api/Features.hs

```diff
@@ -0,0 +1,54 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module Cardano.Api.Features
+  ( Feature(..)
+  , SupportedInEra(..)
+  , supportedInEra
+  ) where
+
+import           Cardano.Api.Eras
+import           Data.Aeson (ToJSON (..))
+
+data ProtocolParameterUTxOCostPerWord
+
+data ProtocolParameterUTxOCostPerByte
+
+-- | A representation of a feature that is supported in a given era.
+data Feature f where
+  ProtocolParameterUTxOCostPerWord :: Feature ProtocolParameterUTxOCostPerWord
+  ProtocolParameterUTxOCostPerByte :: Feature ProtocolParameterUTxOCostPerByte
+
+deriving instance Eq (Feature f)
+deriving instance Show (Feature f)
+
+instance ToJSON (Feature f) where
+  toJSON = toJSON . show
+
+-- | A representation of a feature whether a feature is supported in a given era.
+data SupportedInEra f era where
+  ProtocolParameterUTxOCostPerWordSupportedInAlonzoEra  :: SupportedInEra ProtocolParameterUTxOCostPerWord AlonzoEra
+
+  ProtocolParameterUTxOCostPerByteSupportedInBabbageEra :: SupportedInEra ProtocolParameterUTxOCostPerByte BabbageEra
+  ProtocolParameterUTxOCostPerByteSupportedInConwayEra  :: SupportedInEra ProtocolParameterUTxOCostPerByte ConwayEra
```

Comment:
The `SupportedInEra` type allows us to only add code for when a feature is supported in an era.  We don't have to add the negative cases.

Reply:
Why do we need this type? It's a just a cartesian product of eras and features in a form of a constructor for each, so it will be cumbersome to maintain I think. Couldn't just returning `True` /`False` work in `supportedInEra`?

Reply:
It's a just a cartesian product of eras and features in a form of a constructor for each, so it will be cumbersome to maintain I think. How about just a type class?
```haskell
data ByronEra
data ShelleyEra
data AllegraEra
data MaryEra
data AlonzoEra
data BabbageEra
data ConwayEra

-- This can be turned into a single datatype using DataKinds
data ProtocolParameterUTxOCostPerWord
data ProtocolParameterUTxOCostPerByte
data UnknownFeature

data CardanoEra era where
     ByronEra   :: CardanoEra ByronEra
     ShelleyEra :: CardanoEra ShelleyEra
     AllegraEra :: CardanoEra AllegraEra
     MaryEra    :: CardanoEra MaryEra
     AlonzoEra  :: CardanoEra AlonzoEra
     BabbageEra :: CardanoEra BabbageEra
     ConwayEra  :: CardanoEra ConwayEra


data Feature f where
  ProtocolParameterUTxOCostPerWord :: Feature ProtocolParameterUTxOCostPerWord
  ProtocolParameterUTxOCostPerByte :: Feature ProtocolParameterUTxOCostPerByte
  UnknownFeature :: Feature UnknownFeature

class SupportedInEra ft e where
  supportedInEra :: ft -> CardanoEra e -> Bool

-- default value
instance {-# OVERLAPPABLE #-} SupportedInEra (Feature f) e where
  supportedInEra _ _ = False

instance SupportedInEra (Feature ProtocolParameterUTxOCostPerWord) ByronEra where
  supportedInEra _ _ = True

instance SupportedInEra (Feature ProtocolParameterUTxOCostPerByte) ByronEra where
  supportedInEra _ _ = True

-- >>> supportedInEra ProtocolParameterUTxOCostPerWord ByronEra
-- True

-- >>> supportedInEra ProtocolParameterUTxOCostPerByte ByronEra
-- True

-- >>> supportedInEra UnknownFeature ByronEra
-- False

-- This can be simplified further with a phantom type, not requiring to create a new constructor for each
-- feature, which is already a type
data Feature' f = Feature'

instance {-# OVERLAPPABLE #-} SupportedInEra (Feature' f) e where
  supportedInEra _ _ = False

instance SupportedInEra (Feature' ProtocolParameterUTxOCostPerWord) ByronEra where
  supportedInEra _ _ = True

instance SupportedInEra (Feature' ProtocolParameterUTxOCostPerByte) ByronEra where
  supportedInEra _ _ = True


-- >>> supportedInEra (Feature' @ProtocolParameterUTxOCostPerByte) ByronEra
-- True

-- >>> supportedInEra (Feature' @ProtocolParameterUTxOCostPerByte) ByronEra
-- False

-- >>> supportedInEra (Feature' @UnknownFeature) ByronEra
-- False
```